compiler/K_freevars.fx

```diff
@@ -0,0 +1,285 @@
+    converting each mutable free variable to immutable free variable and 
+    set of mutable usual variables. It create reference variable with value of
+    original freevar and dereferencing it into local temporary variables(tempref) 
+    in start of each function using original freevar.
+*/
+
+from Ast import *
+from K_form import *
+import K_lift_simple
+import Hashmap, Hashset
+
+type fv_func_info_t =
+{
+    fv_fvars: id_hashset_t;
+    fv_declared_inside: id_hashset_t;
+    fv_called_funcs: id_hashset_t
+}
+
+type fv_env_t = (id_t, fv_func_info_t) Hashmap.t
+
+fun empty_fv_env(size0: int){
+    val fv_info0 = fv_func_info_t {
+        fv_fvars = empty_id_hashset(1),
+        fv_declared_inside = empty_id_hashset(1),
+        fv_called_funcs = empty_id_hashset(1) }
+
+    Hashmap.empty(size0, noid, fv_info0)    
+}
+
+fun collect_free_vars(kmods: kmodule_t list, globals: id_t Hashset.t,
+                                                collect_function_info:bool, 
+                                                only_mutable:bool, 
+                                                transitive_closure:bool): fv_env_t {
+// collect_function_info - don't collect fv_called_funcs info if it's false.
+// only_mutable - filter out immutable varibles
+// transitive_closure - compute the transitive closure of the free var sets,
+//     i.e. all the free variables of each function,
+//     directly accessed or via other functions that the function calls.
+//     The term 'closure' is used here with a different meaning,
+//     it's not a function closure but rather a transitive closure of the set.
+    var fv_env : fv_env_t = empty_fv_env(1024)
+
+    fun fold_fv0_ktyp_(t: ktyp_t, loc: loc_t, callb: k_fold_callb_t) {}
+    fun fold_fv0_kexp_(e: kexp_t, callb: k_fold_callb_t) {
+        fold_kexp(e, callb)
+        match e {
+        | KDefFun (ref {kf_name, kf_loc}) =>
+            if !globals.mem(kf_name) {
+                val uv = used_by(e::[], 256)
+                val dv = declared(e::[], 256)
+                // from the set of free variables we exclude global functions, values and types
+                // because they are not included into a closure anyway
+                val called_funcs = empty_id_hashset(16)
+                if(collect_function_info){
+                    uv.app(fun (n) {
+                        match kinfo_(n, kf_loc) {
+                        | KFun _ => if !globals.mem(n) { called_funcs.add(n) }
+                        | _ => {}
+                        }})
+                }
+                val fv0 = empty_id_hashset(uv.size())
+                uv.app(fun (fv) {
+                    match kinfo_(fv, kf_loc) {
+                    | KVal _ =>
+                        if !dv.mem(fv) && !globals.mem(fv) {
+                            if(!only_mutable || is_mutable(fv, get_idk_loc(fv, noloc))){  //TODO: Does already-know-result-thus-stop-calculation rule work in ficus?
```

Comment:
yes, `&&` and `||` are calculated lazily, the second part is only calculated if the first part does not give an asnwer